torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1979,6 +1983,906 @@ def clip_grad_norm_(
+                starting from zero and need to all be increased by an offset
+                for the overall optimizer state dict.
+        """
+        assert sum(p is flat_param for p in fsdp_module.params) == 1, \
+            "`fsdp_module` must own `flat_param`"
+        tensor_state, non_tensor_state, zero_dim_tensor_state, pad_per_rank = \
+            fsdp_module._communicate_optim_state(
+                flat_param, flat_param_state, target_rank,
+            )
+        unflat_param_state = fsdp_module._unflatten_communicated_optim_state(
+            flat_param,
+            tensor_state,
+            non_tensor_state,
+            zero_dim_tensor_state,
+            pad_per_rank,
+        )
+        return unflat_param_state
+
+    def _communicate_optim_state(
+        self,
+        flat_param: FlatParameter,
+        flat_param_state: Dict[str, Any],
+        target_rank: int = 0,
+    ) -> Tuple[Dict[str, List], Dict[str, List], Dict[str, List], List[int]]:
+        """
+        Communicates the optimizer state for a flattened parameter
+        ``flat_param`` across ranks so that the target rank ``target_rank``
+        holds the entire non-sharded optimizer state.
+
+        If ``N`` is the number of tensor optimizer states in the optimizer
+        state dict, then the communication complexity is 0 if ``N = 0`` and
+        ``N + 1`` otherwise (where the plus 1 comes from all-gathering the
+        padding per rank).
+
+        Args:
+            flat_param (FlatParameter): The flattened parameter.
+            flat_param_state (Dict[str, Any]): The entry in the "state" part of
+            the optimizer state dict corresponding to the flattened parameter.
+            target_rank (int): Rank on which to consolidate the optimizer
+                state. (Default: 0)
+
+        Returns:
+            tensor_state (Dict[str, List]): Mapping from positive-dimension
+                tensor state names to :class:`list` s of positive-dimension
+                tensor state values with length equal to the world size.
+            non_tensor_state (Dict[str, List]): Mapping from non-tensor state
+                names to :class:`list` s of non-tensor state values with length
+                equal to the world size.
+            zero_dim_tensor_state (Dict[str, List]): Mapping from zero-
+                dimension tensor state names to :class:`list` s of zero-
+                dimension tensor state values with length equal to the world
+                size.
+            pad_per_rank (List[int]): The flattened parameter's padding per
+                rank in bytes if the parameter has any tensor optimizer state;
+                an empty :class:`list` otherwise.
+        """
+        param_index = -1
+        for i, param in enumerate(self.params):
+            if param is flat_param:
+                param_index = i
+                break
+        assert param_index >= 0, "The FSDP instance must own `flat_param`"
```

Comment:
When I initially wrote the code, I tried to avoid assuming that there was only a single flattened parameter per FSDP module. I can check again if there are any places that assume this.

The reason why I think the assumption is not being made is that each one of these flatten/unflatten functions is **per-parameter**, which means that it will be called on each flattened parameter, and I explicitly construct the mapping from FSDP flattened parameters to their owning modules (instead of assuming that the `i`th flattened parameter is owned by the `i`th FSDP instance like in Fairscale).

Reply:
Oh I see, that makes sense, here is to search the index of flat_param. It was my bad I understood wrong